src/elm/external.ts

```diff
@@ -22,21 +25,48 @@ export class Retrieve extends Expression {
   }
 
   exec(ctx: Context) {
-    let records = ctx.findRecords(this.templateId != null ? this.templateId : this.datatype);
-    let codes = this.codes;
-    if (this.codes && typeof this.codes.exec === 'function') {
-      codes = this.codes.execute(ctx);
-      if (codes == null) {
+    // Object with retrieve information to pass back to patient source
+    // Always assign datatype. Assign codeProperty and dateProperty if present
+    const retrieveDetails: RetrieveDetails = {
+      datatype: this.datatype,
+      ...(this.codeProperty ? { codeProperty: this.codeProperty } : {}),
+      ...(this.dateProperty ? { dateProperty: this.dateProperty } : {})
+    };
+
+    if (this.codes) {
+      const resolvedCodes: Code[] | ValueSet | undefined = this.codes.execute(ctx);
+
+      if (resolvedCodes == null) {
         return [];
       }
+
+      retrieveDetails.codes = resolvedCodes;
     }
-    if (codes) {
-      records = records.filter((r: any) => this.recordMatchesCodesOrVS(r, codes));
-    }
+
     // TODO: Added @dateProperty check due to previous fix in cql4browsers in cql_qdm_patient_api hash: ddbc57
-    if (this.dateRange && this.dateProperty) {
-      const range = this.dateRange.execute(ctx);
-      records = records.filter((r: any) => range.includes(r.getDateOrInterval(this.dateProperty)));
+    if (this.dateRange) {
```

Comment:
Oops.  I just read the TODO on line 46 for the first time.  I told you to remove the `dateProperty` check, but the TODO suggests it should be there.  But I think that's because of it's (prior) use in the filter.  You still check for it before filtering below, so I think it's probably OK that we don't check for it here.

I just went through the history and tracked this back to June of 2017 (courtesy of JB).  ;-)

Anyway, I still don't think we need the dateProperty check here, do you?  If not, then I suggest removing the confusing TODO.

Reply:
Ah, sorry, it's not showing the comment in the context above (on the Conversation page).  Here it is:
```ts
// TODO: Added @dateProperty check due to previous fix in cql4browsers in cql_qdm_patient_api hash: ddbc57
```

Reply:
agreed